agent/util-scripts/pbench-tool-data-sink

```diff
@@ -0,0 +1,733 @@
+    sent from tool meisters via an HTTP PUT method.
+    """
+
+    class Terminate(Exception):
+        pass
+
+    def __init__(self, redis_server, channel, benchmark_run_dir, logger):
+        super(ToolDataSink, self).__init__()
+        # Save external state
+        self.redis_server = redis_server
+        self.channel = channel
+        self.benchmark_run_dir = benchmark_run_dir
+        self.logger = logger
+        # Initialize internal state
+        self._hostname = os.environ["full_hostname"]
+        self.state = None
+        self.tool_data_ctx = None
+        self.directory = None
+        self._data = None
+        self._tm_tracking = None
+        self._lock = Lock()
+        self._cv = Condition(lock=self._lock)
+        # Setup the Bottle server route and the WSGI server instance.
+        self.route(
+            "/tool-data/<tool_data_ctx>/<hostname>",
+            method="PUT",
+            callback=self.put_document,
+        )
+        self._server = DataSinkWsgiServer(host="0.0.0.0", port=8080, logger=logger)
+        # Setup the Redis server channel subscription
+        logger.debug("pubsub")
+        self._pubsub = redis_server.pubsub()
+        logger.debug("subscribe %s", channel)
+        self._pubsub.subscribe(channel)
+        logger.debug("listen")
+        self._chan = self._pubsub.listen()
+        # Pull off first message which is an acknowledgement we have
+        # successfully subscribed.
+        logger.debug("next")
+        resp = next(self._chan)
+        assert resp["type"] == "subscribe", f"bad type: {resp!r}"
+        assert resp["pattern"] is None, f"bad pattern: {resp!r}"
+        assert resp["channel"].decode("utf-8") == channel, f"bad channel: {resp!r}"
+        assert resp["data"] == 1, f"bad data: {resp!r}"
+        logger.debug("next success")
+        # Tell the entity that started us who we are indicating we're ready.
+        started_msg = dict(kind="ds", hostname=self._hostname, pid=os.getpid())
+        logger.debug("publish *-start")
+        redis_server.publish(
+            f"{channel}-start", json.dumps(started_msg, sort_keys=True)
+        )
+        logger.debug("published *-start")
+        self.web_server_thread = None
+
+    def run(self):
+        """run - Start the Bottle web server running and the watcher thread.
+        """
+        self.logger.info("Running Bottle web server ...")
+        try:
+            super().run(server=self._server)
+        except Exception:
+            self.logger.exception("Exception encountered in Bottle web server")
+        finally:
+            self.logger.info("Bottle web server exited")
+
+    def execute(self):
+        """run - Start the Bottle web server running and the watcher thread.
+        """
+        self.web_server_thread = Thread(target=self.run)
+        self.web_server_thread.start()
+        self.logger.debug("web server 'run' thread started")
+
+        try:
+            self.logger.debug("for payload in _chan: ...")
```

Comment:
This doesn't seem useful since we just logged the web server start immediately before, and we're logging each iteration of the `for` below. (Although in extreme cases there might be some value in logging the payloads we're going to process.)

Reply:
These added debug statements were from a desperate attempt to figure out what was happening when the Redis client was getting hosed by the `DaemonContext`.  I didn't take them out because they helped me zero in on what was happening, but there are a lot of them.

I have fixed up these two to combine them, and then tweaked others, and added one that was missing.